File: Recoil/Recoil/Recoil.swift
//
//  Recoil.swift
//  Recoil
//
//  Created by Leland Richardson on 12/12/17.
//  Copyright © 2017 Leland Richardson. All rights reserved.
//

import Foundation

func isRunningInTestEnvironment() -> Bool {
  return true
}

/*
 BIG TODOS:

 - refs
 - proper handling of text children...
 - proper handling of composed Text - View -> Text components
 - vsync batching for setState
 - potentially throw all this work away and reimplement to mimic Fiber's architecture!
 - Animated API?
 - scrollview host component
 - touchable host component
 - textinput host component
 - better color handling
 - view transforms
 - better text font styling
 */

public class Recoil {
  public static func render(_ element: Element, _ rootView: UIView) {

    if let instance = rootView.recoilRoot?.rootInstance {
      // update
      updateInstance(element, rootView, instance)
    } else {
      // mount
      mountInstance(element, rootView)
    }
  }

  public static func unmount(_ rootView: UIView) {
    guard let instance = rootView.recoilRoot?.rootInstance else {
      fatalError()
    }
    unmountInstance(instance, rootView)
  }

  // MARK: private

  private static func updateInstance(_ element: Element, _ rootView: UIView, _ instance: RecoilInstance) {
    if (Reconciler.shouldUpdateComponent(prev: instance.currentElement, next: element)) {
      instance.updateComponent(from: instance.currentElement, to: element)
    } else {
      // Unmount and then mount the new one
      unmountInstance(instance, rootView)
      mountInstance(element, rootView)
    }
  }

  private static func mountInstance(_ element: Element, _ rootView: UIView) {

    let root = RecoilRoot(rootView: rootView)

    rootView.recoilRoot = root

    let instance = Reconciler.instantiateComponent(element: element, root: root)

    root.rootInstance = instance

    let view = Reconciler.mountComponent(instance: instance)

    rootView.removeAllSubviews()

    if let view = view {
      rootView.addSubview(view)
    }

    // let yoga do its thing
    rootView.yoga.isEnabled = true

    // run layout sync
    rootView.yoga.applyLayout(preservingOrigin: false)
  }

  private static func unmountInstance(_ instance: RecoilInstance, _ rootView: UIView) {
    Reconciler.unmountComponent(instance: instance)
    rootView.removeAllSubviews()
    if let root = rootView.recoilRoot {
      root.rootInstance = nil
      root.rootView = nil
      rootView.recoilRoot = nil
    }
  }
}

File: RecoilAndroidApp/app/src/main/java/com/airbnb/android/recoil/Recoil.kt
package com.airbnb.android.recoil

import android.view.ViewGroup


object Recoil {

  fun render(element: Element, rootView: ViewGroup) {
    val instance = rootView.recoilRoot?.rootInstance
    if (instance != null) {
      updateInstance(element, rootView, instance)
    } else {
      mountInstance(element, rootView)
    }
  }

  fun unmount(rootView: ViewGroup) {
    val instance = rootView.recoilRoot?.rootInstance
    if (instance != null) {
      unmountInstance(instance, rootView)
    } else {
      throw IllegalArgumentException("Must pass in a view with a recoil root")
    }
  }

  private fun mountInstance(element: Element, rootView: ViewGroup) {
    val root = RecoilRoot(rootView)

    rootView.recoilRoot = root

    val instance = Reconciler.instantiateComponent(element, root)

    root.rootInstance = instance

    val view = Reconciler.mountComponent(instance)

    rootView.removeAllViews()

    if (view != null) {
      rootView.addView(view)
    }

    // let yoga do its thing
//    rootView.yoga.isEnabled = true

    // run layout sync
//    rootView.yoga.applyLayout(preservingOrigin: false)
  }

  private fun updateInstance(element: Element, rootView: ViewGroup, instance: RecoilInstance) {

  }

  private fun unmountInstance(instance: RecoilInstance, rootView: ViewGroup) {

  }

}
